src/main/scala/uk/gov/nationalarchives/tdr/api/service/TransferAgreementService.scala

```diff
@@ -1,23 +1,32 @@
 package uk.gov.nationalarchives.tdr.api.service
 
+import uk.gov.nationalarchives.Tables
+
 import java.sql.{SQLException, Timestamp}
 import java.util.UUID
-
 import uk.gov.nationalarchives.Tables._
-import uk.gov.nationalarchives.tdr.api.db.repository.ConsignmentMetadataRepository
+import uk.gov.nationalarchives.tdr.api.db.repository.{ConsignmentMetadataRepository, ConsignmentStatusRepository}
 import uk.gov.nationalarchives.tdr.api.graphql.DataExceptions.InputDataException
 import uk.gov.nationalarchives.tdr.api.graphql.fields.TransferAgreementFields.{AddTransferAgreementInput, TransferAgreement}
 import uk.gov.nationalarchives.tdr.api.service.TransferAgreementService._
 
 import scala.concurrent.{ExecutionContext, Future}
 
-class TransferAgreementService(consignmentMetadataRepository: ConsignmentMetadataRepository, uuidSource: UUIDSource, timeSource: TimeSource)
-                              (implicit val executionContext: ExecutionContext) {
+class TransferAgreementService(consignmentMetadataRepository: ConsignmentMetadataRepository,
+                               consignmentStatusRepository: ConsignmentStatusRepository,
+                               uuidSource: UUIDSource, timeSource: TimeSource) (implicit val executionContext: ExecutionContext) {
 
   def addTransferAgreement(input: AddTransferAgreementInput, userId: UUID): Future[TransferAgreement] = {
-    consignmentMetadataRepository.addConsignmentMetadata(convertInputToPropertyRows(input, userId)).map(
+    val transferAgreement = consignmentMetadataRepository.addConsignmentMetadata(convertInputToPropertyRows(input, userId)).map(
       rows => convertDbRowsToTransferAgreement(input.consignmentId, rows)
     )
+    addTransferAgreementStatus(input.consignmentId, "Completed")
```

Comment:
The result of this method call isn't assigned to anything. Because it's a Future, it's eager and it's still executed but because you're not mapping or flat mapping, you might have problems getting any errors back to the client and you might have problems if the database is slow and the Future doesn't complete before Akka returns and kills the thread.

I like the for comprehension so you can do it like this:
```scala
for {
      transferAgreement <- consignmentMetadataRepository.addConsignmentMetadata(convertInputToPropertyRows(input, userId)).map(
        rows => convertDbRowsToTransferAgreement(input.consignmentId, rows)
      )
      _ <- addTransferAgreementStatus(input.consignmentId, "Completed")
    } yield transferAgreement
```
If you don't need the for comprehension then you can use map and flatMap
```scala
    consignmentMetadataRepository.addConsignmentMetadata(convertInputToPropertyRows(input, userId)).map(
      rows => convertDbRowsToTransferAgreement(input.consignmentId, rows)
    ).flatMap(transferAgreement => {
      addTransferAgreementStatus(input.consignmentId, "Completed").map(_ => transferAgreement)
    })
```
For either one, you're waiting for the Future to complete or fail before the `addTransferAgreement` method returns.
This is the main reason that Future is a pain to work with :smile:

Reply:
Ah, of course; this makes sense. Thanks, I'll add the for comprehension